docs/testing.md

```diff
@@ -0,0 +1,81 @@
+
+## General guidelines
+
+- Imports should be alphabetized, even in test files.
+- Comments should be full sentences to ease readability.
+- There are a lot of helpers in the `tests/unit/helpers` and `tests/unit/amo/helpers` modules–please use them.
+- Use action creators in `amo/actions/` and `core/actions/` instead of hard-coding `dispatch()` arguments or state data for tests. This applies to both UI components and reducers/sagas.
+- Use constants (see `core/constants`) when using the same value across files. This avoids hard-coding values and [magic constants](https://en.wikipedia.org/wiki/Magic_constant).
+
+
+## Jest
+
+[Jest](https://facebook.github.io/jest/docs/en/getting-started.html) is our main testing framework. Please refer to the [`README` section about running the test suite](https://github.com/mozilla/addons-frontend#running-tests) to know how to run the tests. Below are a few rules regarding Jest:
+
+- Prefer `toEqual()` over `toBe()`.
+- Do not use `expect.assertions(N)`, this is hard to maintain. Instead, add a `catch()` branch to detect unexpected errors.
+
+When creating a new test file, start with a `describe()` block that takes the current file name as first argument. This makes easy to find/edit a failing test case as Jest will display the test file in its output:
+
+```js
+describe(__filename, () => {
+});
+```
+
+## Spies/Stubs/Mocks and Sinon.JS
+
+We use [sinon](http://sinonjs.org/) for spies, stubs and mocks. In addition, we use [sinon assertions](http://sinonjs.org/releases/v3.2.1/assertions/) over Jest expectations because failure messages are more descriptive.
+
+```js
+const spy = sinon.spy();
+// ...
+
+// NOT GOOD
+expect(spy.called).toEqual(true);
+
+// GOOD
+sinon.assert.called(spy);
+```
+
+There is no need to import sinon, it is already in the global scope for all test files.
+
+
+## Testing reducers and sagas
+
+For sagas/reducers, there are two useful helpers: `dispatchClientMetadata()` and `dispatchSignInActions()` (`tests/unit/amo/helpers`) that should be used to initialize state in a realistic manner. The former is used to obtain a non-authenticated state while the latter returns an authenticated state.
+
+When you need a `errorHandler` or a `errorHandlerId`, use the `createStubErrorHandler()` helper from `tests/unit/helpers`.
+
+When asserting for exceptions/errors, do not use ES6 shorthand functions/implicit return functions. It explicits what is expected to throw an error:
```

Comment:
Why not? I think this code would have been fine:

```js
expect(() => methodThatThrowsAnError())
  .toThrow(/expected error message/);
```

Reply:
Here's my issue with that (I mentioned this in a previous review somewhere):

Implicit return functions are useful when it's a shorthand for a quick, lambda-like function who return value you want. I don't think implicit-return functions are a good shorthand to use when you aren't testing the return value of a function–in many cases you'd be using them and they wouldn't even have a return value or at least you wouldn't be using it. You're testing a side-effect of calling the function (e.g. throwing an exception) not the value the function returns. But seeing implicit returns in the code makes me scan for a test involving the returned value. So while:

```js
expect(sortAnArray([1, 3, 2])).toEqual([1, 2, 3]);
```
is fine, I don't think:
```js
expect(() => sortAnArray('notAnArray')).toThrow(/Not an array/);
```
is proper as you're using implicit return just to save on typing. You wouldn't write that test as:
```js
expect(() => {
  return sortAnArray('notAnArray');
}).toThrow(/Not an array/);
```
... you'd omit the return. So I don't like using them.

Reply:
This is so much more concise though 😁 

```js
expect(() => sortAnArray('notAnArray')).toThrow(/Not an array/);
```
It makes the test more readable.

Reply:
To be more clear about what I mean. In response to:

> Implicit return functions are useful when it's a shorthand for a quick, lambda-like function [whose] return value you want.

I think arrow functions serve a secondary purpose of conciseness which can really aid in readability.

Reply:
Again though, I don't think it *is* more readable. Seeing an implicit return I read the return value as important but it's just a by-product of a shorter syntax. With the offending/error-throwing code on it's own line, without a return statement, I can see:

1. the return value is not important/relevant to that error being thrown
2. this is code that has a side-effect
3. we're testing the side-effect, as no data is being passed in any direction

I don't think concise === more readable.

Reply:
> Seeing an implicit return I read the return value as important

But you can just look at the context and immediately know that it's not. 

*shrug*

Anyway, this sentence needs fixing:

> It explicits what is expected to throw an error:

I'm not really sure what it meant to say so I don't have a good suggestion.

Reply:
> But you can just look at the context and immediately know that it's not.

While I also think brackets are overkill and it is not too complicated to read, I see a similarity with single-letter variable names here.

Example: `promise.catch((e) => handleError(e))`. Here, the context is self-explanatory I think. `e` is an error, yet writing `error` does not take too much time, and makes things crystal clear.

> I'm not really sure what it meant to say so I don't have a good suggestion.

What I want to say here is that using `() => {}` highlights the function that will throw an error (which is what we expect).

Reply:
I just rephrased the second sentence on this line.